mlflow/pyfunc/__init__.py

```diff
@@ -266,7 +270,117 @@ def _load_model_env(path):
+                if values.dtype == np.object and t not in (DataType.binary, DataType.string):
+                    values = values.infer_objects()
+
+                if values.dtype in (t.to_pandas(), t.to_numpy()):
+                    return values
+
+                if t == DataType.binary and values.dtype.kind == t.binary.to_numpy().kind:
+                    #  NB: bytes in numpy have variable itemsize and would fail the upcast check.
+                    return values
+
+                if t == DataType.string and values.dtype == np.object:
+                    #  NB: strings are by default parsed and inferred as objects, but it is
+                    # recommended to use StringDtype extension type if available. See
+                    #
+                    # `https://pandas.pydata.org/pandas-docs/stable/user_guide/text.html`
+                    #
+                    # for more detail.
+                    try:
+                        return values.astype(t.to_pandas(), errors="raise")
+                    except ValueError:
+                        raise MlflowException(
+                            "Failed to convert column {0} from type {1} to {2}.".format(
+                              name, values.dtype, t)
+                        )
+
+                numpy_type = t.to_numpy()
+                is_compatible_type = values.dtype.kind == numpy_type.kind
+                is_upcast = values.dtype.itemsize <= numpy_type.itemsize
+                if is_compatible_type and is_upcast:
+                    # conversions between compatible types and ints -> floats are allowed.
+                    return values.astype(numpy_type, errors="raise")
+                else:
+                    # NB: conversion between incompatible types (e.g. floats -> ints or
+                    # boolean -> numeric) are not allowed. While supported by pandas and numpy,
+                    # these conversions alter the values significantly.
+                    raise MlflowException("Incompatible input types for column {0}. "
+                                          "Can not safely convert {1} to {2}.".format(name,
+                                                                                      values.dtype,
+                                                                                      numpy_type))
+            col_names = input_schema.column_names()
+            col_types = input_schema.column_types()
+            expected_names = set(input_schema.column_names())
+            actual_names = set(data.columns)
+            missing_cols = expected_names - actual_names
+            extra_columns = actual_names - expected_names
+            if missing_cols:
+                message = "Model input is missing columns {0}.".format(missing_cols)
+                raise MlflowException(message)
+            if extra_columns:
+                _logger.warning("Ignoring unexpected columns %s", extra_columns)
```

Comment:
I don't think we should bother printing this. It could actually be more expensive than doing the actual prediction.

Reply:
ok, makes sense.